zkevm-circuits/src/taiko_pi_circuit.rs

```diff
@@ -0,0 +1,723 @@
+        rlc::value(self.encode_field(idx).iter().rev(), rand)
+    }
+
+    fn keccak_hi_low(&self) -> [F; 2] {
+        let keccaked_pi = keccak256(self.encode_raw());
+        [
+            keccaked_pi
+                .iter()
+                .take(16)
+                .fold(F::ZERO, |acc: F, byte| {
+                    acc * F::from(BYTE_POW_BASE) + F::from(*byte as u64)
+                }),
+                keccaked_pi
+                .iter()
+                .skip(16)
+                .fold(F::ZERO, |acc: F, byte| {
+                    acc * F::from(BYTE_POW_BASE) + F::from(*byte as u64)
+                })
+        ]
+    }
+
+    fn keccak(&self) -> Vec<u8> {
+        keccak256(self.encode_raw()).to_vec()
+    }
+
+    fn keccak_assignment(&self) -> Vec<F> {
+        self.keccak()
+            .iter()
+            .map(|b| F::from(*b as u64))
+            .collect()
+    }
+
+    fn total_len(&self) -> usize {
+        self.encode_raw().len()
+    }
+
+    fn field_len(&self, idx: usize) -> usize {
+        self.encode_field(idx).len()
+    }
+
+
+}
+
+#[derive(Clone, Debug)]
+pub struct TaikoPiCircuitConfig<F: Field> {
+    q_enable: Selector,
+    keccak_instance: Column<Instance>, // equality
+
+    meta_hash: FieldGadget<F>,
+    parent_hash: (Cell<F>, FieldGadget<F>, Cell<F>),
+    block_hash: (Cell<F>, FieldGadget<F>, Cell<F>),
+    signal_root: FieldGadget<F>,
+    graffiti: FieldGadget<F>,
+    prover: FieldGadget<F>,
+
+    total_acc: Cell<F>,
+    keccak_bytes: FieldGadget<F>,
+    keccak_hi_lo: [Cell<F>; 2],
+
+    block_table: BlockTable,
+    keccak_table: KeccakTable,
+    byte_table: ByteTable,    
+
+    annotation_configs: Vec<CellColumn<F, PiCellType>>,
+}
+
+pub struct TaikoPiCircuitConfigArgs<F: Field> {
+    /// 
+    pub evidence: PublicData<F>,
+    /// BlockTable
+    pub block_table: BlockTable,
+    /// KeccakTable
+    pub keccak_table: KeccakTable,
+    /// ByteTable
+    pub byte_table: ByteTable,
+    /// Challenges
+    pub challenges: Challenges<Expression<F>>,
+}
+
+
+impl<F: Field> SubCircuitConfig<F> for TaikoPiCircuitConfig<F> {
+    type ConfigArgs = TaikoPiCircuitConfigArgs<F>;
+    /// Return a new TaikoPiCircuitConfig
+    fn new(
+        meta: &mut ConstraintSystem<F>,
+        Self::ConfigArgs {
+            evidence,
+            block_table,
+            keccak_table,
+            byte_table,
+            challenges,
+        }: Self::ConfigArgs,
+    ) -> Self {
+        let keccak_r = challenges.keccak_input();
+        let evm_word = challenges.evm_word();
+        let cm = CellManager::new(
+            meta,
+            vec![
+                (PiCellType::Byte, 7, 1, false),
+                (PiCellType::StoragePhase1, 1, 1, true),
+                (PiCellType::StoragePhase2, 1, 1, true),
+            ],
+            0,
+            DEFAULT_LEN,
+        );
+        let mut cb: ConstraintBuilder<F, PiCellType> = ConstraintBuilder::new(4,  Some(cm.clone()), Some(evm_word.expr()));
```

Comment:
Is this set to `4` instead of `5` because the `q_enable` selector is added afterwards and so doesn't get counted while adding the constraints? Seems like another good reason not to do it I think, more edge cases to take into account! (no problem of doing it like this for this temporary version though).

Reply:
Ahhh yea! you're right this is such a import point I didn't thought of. I'll revert the change to `cb.build_constraints` then